File: examples/pytest/navigation_test.py
from pytest import mark


@mark.describe("Back navigation")
@mark.xfail(reason="https://github.com/alumnium-hq/alumnium/pull/154")
def test_navigate_back_uses_history(al, navigate):
    # Arrange: create history with two pages
    navigate("https://example.com")
    assert "example.com" in al.driver.url

    al.do("Click on More information")
    assert "www.iana.org" in al.driver.url

    # Act: ask the agent to go back to the previous page
    al.do("Navigate back to the previous page")

    # Assert: we are back on the first page
    assert "example.com" in al.driver.url


@mark.describe("Back navigation - different phrases")
@mark.xfail(reason="https://github.com/alumnium-hq/alumnium/pull/154")
def test_back_navigation_different_phrases(al, navigate):
    """Test that different ways of asking to go back work without explicit learning examples."""
    # Arrange: create history with two pages
    navigate("https://example.com")
    assert "example.com" in al.driver.url

    al.do("Click on More information")
    assert "www.iana.org" in al.driver.url

    # Test different ways to ask for back navigation
    test_phrases = ["Go back", "Navigate back", "Return to the previous page"]

    for phrase in test_phrases:
        # Go forward again to test going back
        if "example.com" in al.driver.url:
            al.do("Click on More information")
            assert "www.iana.org" in al.driver.url

        # Test the back navigation phrase
        al.do(phrase)
        assert "example.com" in al.driver.url, f"Failed with phrase: '{phrase}'"

File: packages/typescript/examples/mocha/navigation.test.ts
import { Alumni, Model, NavigateBackTool } from "alumnium";
import assert from "assert";
import "./globals.js";
import { navigate } from "./helpers.js";

describe("Navigation", () => {
  const shouldSkip = () => {
    if (Model.current.provider === "mistralai") {
      return "Needs more work";
    }
    return null;
  };

  it("navigate back uses history", async function () {
    if (shouldSkip()) {
      this.skip();
    }

    const alWithNavBack = new Alumni(driver, {
      extraTools: [NavigateBackTool],
    });

    await navigate(driver, "https://the-internet.herokuapp.com");
    assert.strictEqual(
      await alWithNavBack.driver.url(),
      "https://the-internet.herokuapp.com/",
    );

    await alWithNavBack.do("open typos");
    assert.strictEqual(
      await alWithNavBack.driver.url(),
      "https://the-internet.herokuapp.com/typos",
    );

    await alWithNavBack.do("navigate back to the previous page");
    assert.strictEqual(
      await alWithNavBack.driver.url(),
      "https://the-internet.herokuapp.com/",
    );
  });
});
